src/sensors/lorreconstruction.rs

```diff
@@ -79,18 +104,44 @@ fn dot_product_clusters(hits: &[MCQT], xyzs: &SensorMap, threshold: f32, min_sen
     if c1.len() >= min_sensors && c2.len() >= min_sensors { Some((c1, c2)) } else { None }
 }
 
-/// Reconstruct LORs based on the barycentre of the 2 selected clusters (dot prod) without DOI.
-fn lor_from_sensor_positions(hits       : &[MCQT]   ,
-                             xyzs       : &SensorMap,
-                             threshold  : f32       ,
-                             min_sensors: usize     ,
-                             min_charge : f32
-                            ) -> Option<Hdf5Lor> {
+// Is there a more generic way to avoid rewriting this code over and over again?
+fn get_sensor_charge(hits: &[MCQT]) -> Vec<Ratio> {
+    let mut qs = Vec::with_capacity(hits.len());
+    for &MCQT { q, .. } in hits {
+        qs.push(ratio(q as f32));
+    }
+    qs
+}
```

Comment:
Idiomatically, this would be written approximately like this:
```rust
fn get_sensor_charge(hits: &[MCQT]) -> Vec<Ratio> {
    hits.iter()
        .map(|&MCQT { q, .. }| ratio(q as f32))
        .collect()
}
```
I guess that by "rewriting this code over and over again" you mean the pattern of mapping an iterable into a `Vec`?

I'm sure that there is. Whether it is worth the hassle introduced by dependence on whether you want to clone, copy or extract from refs, is more doubtful.

Reply:
I tried basically what you suggest but was getting compile errors, could be that I had some other error with a knock on effect though. I'll update to the neater form.

My comment was because I seemed to be repeating the same lines of code multiple times to get a vector of the charge or the sum or the same for x or y.... I looked if it was possible to have a function like `get_property_vec` or something using a variable that selects the parameter you want but I understood that it's compile time unstable so I gave up.

Reply:
Probably not worth it for just two of these, but if you find yourself doing it more often, then maybe something along these lines might make sense

```rust
macro_rules! xxx {
    (for $name:ident in $collection:ident collect $extract:expr) => {
        $collection.iter()
            .map(|&$name| $extract)
            .collect()
    };
}

fn get_sensor_charge(hits: &[SensorReadout]) -> Vec<Ratio>  { xxx!(for h in hits collect ratio(h.q as f32)) }
fn get_sensor_z     (hits: &[SensorReadout]) -> Vec<Length> { xxx!(for h in hits collect h.z) }
```

Reply:
Bah!

My original macro was to be used like this: `xxx!(hits, h, h.z)` This was too cryptic, so I changed it to the more verbose version I posted above, without noticing that, by the time you give it a meaningful name, there is very little reduction in length/noise, and it becomes completely pointless.

Maybe, if you give it a good name, it can be used without being wrapped in functions. Perhaps like this:
```rust
let zs: Vec<_> = collect!(for h in hits get h.z);
let rs: Vec<_> = collect!(for h in hits get (h.x*h.x + h.y*h.y).sqrt());
let sensor_charge: HashMap<_, _> = collect!(for h in hits get (h.sensor_id, h.q));
```

Anyway, the point is that in these kinds of situations, marcos might give you better mileage than doing battle with type bounds in generics.